packages/file/lib/src/backends/memory/memory_file.dart

```diff
@@ -416,19 +416,18 @@ class _FileSink implements io.IOSink {
   Future<void> addStream(Stream<List<int>> stream) {
     _checkNotStreaming();
     _streamCompleter = Completer<void>();
-    void finish() {
-      _streamCompleter!.complete();
-      _streamCompleter = null;
-    }
 
     stream.listen(
       (List<int> data) => _addData(data),
       cancelOnError: true,
       onError: (Object error, StackTrace stackTrace) {
-        _completer.completeError(error, stackTrace);
-        finish();
+        _streamCompleter!.completeError(error, stackTrace);
```

Comment:
Actually, just to be safe these should probably be `?.`s? I think it might be possible to get both an error on a stream as well as a done event?

Reply:
Not with `cancelOnError`:

https://api.flutter.dev/flutter/dart-async/Stream/listen.html#:~:text=If%20cancelOnError%20is,functions%20are%20called.

More specifically, it's only possible if the stream implementation doesn't respect the subscription cancel, which is the kind of error that should probably throw exceptions left and right anyway.

Reply:
Although, I have seen (and seconded) some contention on the wording of this contract, while not in `cancelOnError`, on the presumably analogous but possibly asynchronous `StreamSubscription.cancel`:

https://github.com/flutter/flutter/issues/91814

Do you happen to have a case on hand that sends a done event after the `cancel` has started? I can look around to see if I have one somewhere, but it may not exist.

Reply:
I don't have a specific situation or sequence of events in mind :). Good point about `cancelOnError`, although I don't love relying on that contract (also somebody could easily remove that later, and not know that these now need to be a `?.`).

It just seems like good defensive programming to me to use `?.` here, and there is little downside (other than swallowing some errors, which isn't great, but the alternative is crashing).

Reply:
I made the mistake of looking at https://github.com/dart-lang/sdk/blob/main/sdk/lib/async/stream_impl.dart and now I don't know how any of it works at all.

I typically prefer crashes over swallowed errors, but I'm also quick to change my tune when it hits the UI. On the other hand, this PR would not introduce any crashes that wouldn't already have been happening, since the code was previously calling `finish()` anyway, which also accesses `_streamCompleter` and sets it to null. From that perspective, adding the defensive `?.` would constitute and additional change in behavior I'd prefer to avoid in this PR.

Reply:
I'm at least going to go ahead and add the test case so that if someone does change the SDK stream impl, we have a failing unit test to let people know that something needs to be migrated.

Reply:
Ok, that sgtm and its true that your PR isn't really changing this.